### DISCUSSION TITLE
Onigurumi Regex to match a JSON block?

### DISCUSSION THREAD

**Discussion by JustinGrote** [microsoft/vscode-discussions#808] [Extension Development QnA] [answered]
I'm making a new log file highlighter (Crisp Log Highlighter) and the embedded language is pretty neat. However I'm having a hard time coming up with a good onigurumi compatible regex that will match a JSON block so I can use that syntax highlighting instead. Does anyone have a good example? Has to support nested braces, which is the trickiest part which mostly seems to need a recursive lookup.

**Comment by theReynald** [comment, marked as answer]
Hi justingrote,

Creating an Oniguruma-compatible regular expression for matching JSON blocks, especially those with nested braces, can be challenging due to the recursive nature of JSON. Here's a regex pattern and explanation that might help you get started:

```regex
\{(?:[^{}]|(?R))*\}
```

Breaks down like this:

- `\{` matches the opening curly brace `{`.
- `(?: ... )` is a non-capturing group that allows us to group elements together.
- `[^{}]` matches any character that is not an opening or closing curly brace.
- `|` is the OR operator.
- `(?R)` is a recursive reference to the entire pattern.

Here's how it basically works:

1. `\{` matches the opening curly brace of the JSON block.
2. `(?: ... )*` matches any number of characters inside the JSON block, where:
   - `[^{}]` matches any character that is not a curly brace.
   - `(?R)` recursively matches the entire pattern, allowing for nested JSON blocks.
3. `\}` matches the closing curly brace of the JSON block.

This regular expression should work well for simple JSON highlighting and handle nested JSON objects. Keep in mind that JSON can become quite complex, and handling all edge cases with regex alone may not be practical. You may want to consider using a JSON parser to tokenize the input and then apply syntax highlighting based on the tokenized JSON structure. This would provide a more accurate and reliable highlighting for all JSON variations.

Hope this helps.

**Reply by JustinGrote** [reply]
Thank you. In this case it is for a vscode extension grammar parser so tokenization libraries are not an option, but they basically only need to match where the full block begins and ends, so basically as long as the core braces structure gets recursively matched but doesn't generally match other nested-brace style languages, the internal detail is less important.
